### PR TITLE
[optimizer] refactor SGD to use functional API

### PR DESCRIPTION
**Pull request by wanchaol**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #50630 [dist_optim] add warning to distributed optimizer
* #50620 [dist_optim] add distributed functional AdamW optimizer
* #50619 [dist_optim] add distributed functional RMSprop optimizer
* #50623 [dist_optim] add distributed functional Adadelta optimizer
* #50618 [dist_optim] add distributed functional sgd optimizer
* #50624 [dist_optim] add distributed functional Adam optimizer
* #50411 [optimizer] refactor AdamW to use functional API
* #50410 [optimizer] refactor RMSProp to use functional API
* #50409 [optimizer] refactor Adadelta to use functional API
* **#45597 [optimizer] refactor SGD to use functional API**

Differential Revision: [D25932773](https://our.internmc.facebook.com/intern/diff/D25932773)

### CODE REVIEW

**Review by izdeby — APPROVED** [review]

File: torch/optim/sgd.py [hunk 1]
```diff
@@ -86,29 +87,39 @@ def step(self, closure=None):
...
-                d_p = p.grad
-                if weight_decay != 0:
-                    d_p = d_p.add(p, alpha=weight_decay)
-                if momentum != 0:
-                    param_state = self.state[p]
-                    if 'momentum_buffer' not in param_state:
-                        buf = param_state['momentum_buffer'] = torch.clone(d_p).detach()
+                if p.grad is not None:
+                    params_with_grad.append(p)
+                    d_p = p.grad
```
Comment (albanD) [inline torch/optim/sgd.py:102]: nit: you can do directly `d_p_list.append(p.grad)`

File: torch/optim/functional.py [hunk 2]
```diff
@@ -96,3 +96,39 @@ def adam(params: List[Tensor],
...
+
+        d_p = d_p_list[i]
+        if weight_decay != 0:
+            d_p = d_p.add(param, alpha=weight_decay)
+
+        if momentum != 0:
+            buf = momentum_buffer_list[i]
+
+            if buf is None:
+                buf = torch.clone(d_p).detach()
```
Comment (albanD) [inline torch/optim/functional.py:124]: nit: why the detach here? Are there any case where this function is called outside of a `torch.no_grad()` block?

Comment (wanchaol) [reply torch/optim/functional.py:124]: not really, it should all under a `torch.no_grad()` block, I just copied this from the sgd.py, if there's no need to detach here, I can remove it.

Comment (albanD) [reply torch/optim/functional.py:124]: Ho I didn't realize it was in the original version. I think you should keep it as is here to reduce any chance to break stuff in this refactor. :)

**Review by albanD — APPROVED** [review]
LGTM !
